### ISSUE TITLE
Peer to peer remote sharing of viewer + data

### ISSUE THREAD

**Issue by jni** [napari/napari#1875]
This has been on our wishlist for some time. A few days ago I came across xpublish on Twitter, which seems like something that would make this relatively easy!

https://medium.com/pangeo/xpublish-ff788f900bbf

https://github.com/xarray-contrib/xpublish

xref #881 

Just jotting this down somewhere shared, I haven't had time to play with it yet. Use cases:

- someone has a large dataset that is a pain to share and just wants to allow a collaborator to look at the data directly, rather than with screen sharing
- collaborative annotation of a volume

I'd see this working similar to VSCode Live Share: just send someone a link and they can open up their napari but pointed at your data.

**Comment by sofroniewn** [comment]
Yes, love this idea. I bet @manzt will have thoughts!! I think we're getting close :-)

**Comment by manzt** [comment]
Definitely have some ideas here! I'd been experimenting with being able to do this with a plug-in, but ran into some hiccups with serializing layer state to JSON (like colormaps), and ultimately zarr.

Love the idea of being able to turn your viewer into a server. Whether that lives within napari or a plug-in.


```python
viewer.serve() # or napari_server_plugin.serve(viewer)
```

**Comment by kevinyamauchi** [comment]
@manzt, I also love the idea of being able to turn the viewer into a server. This would be super helpful in my work and it would be great to hear your thoughts when you have time.

> Definitely have some ideas here! I'd been experimenting with being able to do this with a plug-in, but ran into some hiccups with serializing layer state to JSON (like colormaps), and ultimately zarr.

Were you using the `layer._get_state()` method for serialization? I think initially, we had been aiming to make this serializable to JSON, but we may have gotten away from that. Were there particular layers/properties that were giving you trouble (sounds like colormaps were an issue)?

**Comment by manzt** [comment]
> This would be super helpful in my work and it would be great to hear your thoughts when you have time.

Happy to schedule some time to talk about it! It's been a little while since I've looked into this, but sharing some initial ideas (regarding serialization + http server) with you're understanding of napari's internals would likely be fruitful.

> Were you using the layer._get_state() method for serialization?

I was using `layer.as_layer_data_tuple()` (which I'm guessing calls this internally? related issue: https://github.com/napari/napari/issues/1584). 

My initial prototype was to construct an in-memory zarr store (hierarchy) for the viewer, where each layer is a zarr array (or arrays, e.g. pyramid), and then serving that store via HTTP. One could serve arbitrary chunking of the underlying data:


```python
class ViewerStore:
  # ... # 
  def __getitem__(self, key):
    # handle returning .zarray/.zgroup/.zattrs (layer state)
    if key in self._meta_store:
      return self._meta_store[key] # in-memory dict stores all metadata for virtual hierarchy
   
    # Return "chunk" of underlying numpy-like data
    layer, chunk_shape = self._get_layer_and_chunk_shape(key)        # get underlying layer + virtual metadata for layer
    selection = compute_selection_from_chunk_key(key, chunk_shape) # convert chunk key to numpy selection
    bytes = np.asarray(layer.data[selection]).tobytes()
    return bytes
```

Ideally this virtual (in-memory) hierarchy would mirror Nick's work with saving napari to zarr (on disk) https://github.com/napari/napari/issues/490, so that reading the store is identical.

> An important caveat is that I don't think this approach is powerful enough to capture the "Live Share"-like experience that @jni has in mind, which would require syncing layer state between peers.

**Comment by jni** [comment]
> An important caveat is that I don't think this approach is powerful enough to capture the "Live Share"-like experience that @jni has in mind, which would require syncing layer state between peers.

Read-only sharing is good enough for a start!!!

**Comment by kevinyamauchi** [comment]
> Happy to schedule some time to talk about it! It's been a little while since I've looked into this, but sharing some initial ideas (regarding serialization + http server) with you're understanding of napari's internals would likely be fruitful.

That would be great, @manzt ! I'll ping you on the zulip to set up a time. I have a few weeks off in December and this could be a fun project!

> An important caveat is that I don't think this approach is powerful enough to capture the "Live Share"-like experience that @jni has in mind, which would require syncing layer state between peers.

I totally agree with Juan that having read-only sharing would be an amazing first step.
